File: Neetcode/Arrays_and_Hashing/1930_Unique_3_Palindromes.py
from collections import defaultdict

s = "bbcbaba"

# O(k*n) => k is distinct chars
def threePals4(s):
  count = 0
  chars = set(s)
  for char in chars:
      first,last = s.find(char),s.rfind(char)
      count += len(set(s[first+1:last]))
  return count

print(threePals4(s))


# 22 testcases TLE
# O(n^4)
def checkPal(s):
    size = len(s)
    left, right = 0, size - 1
    while (left < right):
        if (s[left] != s[right]):
            return False
        left += 1
        right -= 1
    
    return True
def threePals(s):
    umap = defaultdict(int)
    size = len(s)
    temp = ""
    for i in range(size):
        for j in range(i+1, size):
            for k in range (j+1, size):
                temp += s[i]
                temp += s[j]
                temp += s[k]
                if (checkPal(temp)):
                    umap[temp] += 1
                temp = ""
    return len(umap)

# 44 testcases TLE
# O(n^3)
def threePals2(s):
    umap = defaultdict(int)
    size = len(s)
    temp = ""
    for i in range(size):
        for j in range(i+1, size):
           if s[i] in s[j+1:size]:
                temp += s[i]
                temp += s[j]
                temp += s[i]
                umap[temp] += 1
                temp = ""

    return len(umap)

# 45 testcases TLE
# O(n^2)
def threePals3(s):
    umap = defaultdict(int)
    size = len(s)
    temp = ""
    left, right = 0, size - 1
    while (left < right):
        while (s[right] != s[left] and left < right):
            right -= 1

        new_right = right - 1
        while (left < new_right):
            temp += s[left]
            temp += s[new_right]
            temp += s[right]
            umap[temp] += 1
            temp = ""
            new_right -= 1
        
        new_left = left + 1
        while (s[new_left] == s[left] and new_left < size):
            new_left += 1
          
        left = new_left
        right = size - 1

    return len(umap)         

File: Neetcode/Arrays_and_Hashing/1930_Unique_3_Palindromes.cpp
#include <bits/stdc++.h>
// #define ll long long

using namespace std;

// using simple loops and indices (preferred)
int threePals(string s)
{ 
  int size = s.size();
  unordered_set<char> uset(s.begin(), s.end());
  int count = 0;

  for (auto& it : uset)
  {
    int left = -1, right = -1;

    // find first occurrence
    for (int i = 0; i < size; i++)
    {
      if (s[i] == it)
      {
        left = i;
        break;
      }
    }

    // find last occurrence
    for (int i = size - 1; i >= 0; i--)
    {
      if (s[i] == it)
      {
        right = i;
        break;
      }
    }

    // if first and last occurrence are found and they aren't the same position
    if (left != -1 && right != -1 && left < right)
    { 
      // get count of unique elements in between 
      unordered_set<char> temp;
      for (int i = left + 1; i < right; i++)
      {
        temp.insert(s[i]);
      }
      count += temp.size();
    }
  }

  return count;
}

int main()
{
  string s = "bbcbaba";
  cout << threePals(s);
}

// using iterators
int threePals1(string s)
{
  unordered_set<char> uset(s.begin(), s.end());
  int count = 0;
  for (auto& it : uset)
  {
    auto left = find(s.begin(), s.end(), it);
    auto right = find(s.rbegin(), s.rend(), it);
    if (left != s.end() && right != s.rend() && distance(left, right.base()) > 1)
    {
      unordered_set<char> temp(left + 1, right.base() - 1);
      int uniqSize = temp.size();
      count += uniqSize;
    }
  }

  return count;
}

// using string methods
int threePals2(const std::string& s) {
    int count = 0;

    for (char c : s) {
        std::unordered_set<char> temp;

        for (size_t i = s.find(c) + 1; i < s.rfind(c); ++i) {
            temp.insert(s[i]);
        }

        count += temp.size();
    }

    return count;
}